internal/upstreamldap/upstreamldap.go

```diff
@@ -0,0 +1,386 @@
+	if err != nil {
+		if strings.HasSuffix(err.Error(), ": missing port in address") { // sad to need to do this string compare
+			host = hostAndPort
+			port = defaultPort
+		} else {
+			return "", err // hostAndPort argument was not parsable
+		}
+	}
+	switch {
+	case port != "" && strings.HasPrefix(host, "[") && strings.HasSuffix(host, "]"):
+		// don't add extra square brackets to an IPv6 address that already has them
+		return host + ":" + port, nil
+	case port != "":
+		return net.JoinHostPort(host, port), nil
+	default:
+		return host, nil
+	}
+}
+
+// A name for this upstream provider.
+func (p *Provider) GetName() string {
+	return p.c.Name
+}
+
+// Return a URL which uniquely identifies this LDAP provider, e.g. "ldaps://host.example.com:1234".
+// This URL is not used for connecting to the provider, but rather is used for creating a globally unique user
+// identifier by being combined with the user's UID, since user UIDs are only unique within one provider.
+func (p *Provider) GetURL() string {
+	return fmt.Sprintf("%s://%s", ldapsScheme, p.c.Host)
+}
+
+// TestConnection provides a method for testing the connection and bind settings. It performs a dial and bind
+// and returns any errors that we encountered.
+func (p *Provider) TestConnection(ctx context.Context) error {
+	err := p.validateConfig()
+	if err != nil {
+		return err
+	}
+
+	conn, err := p.dial(ctx)
+	if err != nil {
+		return fmt.Errorf(`error dialing host "%s": %w`, p.c.Host, err)
+	}
+	defer conn.Close()
+
+	err = conn.Bind(p.c.BindUsername, p.c.BindPassword)
+	if err != nil {
+		return fmt.Errorf(`error binding as "%s": %w`, p.c.BindUsername, err)
+	}
+
+	return nil
+}
+
+// DryRunAuthenticateUser provides a method for testing all of the Provider settings in a kind of dry run of
+// authentication for a given end user's username. It runs the same logic as AuthenticateUser except it does
+// not bind as that user, so it does not test their password. It returns the same values that a real call to
+// AuthenticateUser with the correct password would return.
+func (p *Provider) DryRunAuthenticateUser(ctx context.Context, username string) (*authenticator.Response, bool, error) {
+	endUserBindFunc := func(conn Conn, foundUserDN string) error {
+		// Act as if the end user bind always succeeds.
+		return nil
+	}
+	return p.authenticateUserImpl(ctx, username, endUserBindFunc)
+}
+
+// Authenticate an end user and return their mapped username, groups, and UID. Implements authenticators.UserAuthenticator.
+func (p *Provider) AuthenticateUser(ctx context.Context, username, password string) (*authenticator.Response, bool, error) {
+	endUserBindFunc := func(conn Conn, foundUserDN string) error {
+		return conn.Bind(foundUserDN, password)
+	}
+	return p.authenticateUserImpl(ctx, username, endUserBindFunc)
+}
+
+func (p *Provider) authenticateUserImpl(ctx context.Context, username string, bindFunc func(conn Conn, foundUserDN string) error) (*authenticator.Response, bool, error) {
```

Comment:
We probably want something like this to help admins understand if their query is sane:

```go
trace := utiltrace.New("ldap login",
	utiltrace.Field{Key: "name", Value: p.GetName()},
	// maybe more stuff that could be valuable and does not leak PII, username hash?
)
defer trace.LogIfLong(500 * time.Millisecond)
```

Reply:
Added in 3008d1a85cc499bfd5d16f2e7611bf67e07a5f0f